### DISCUSSION TITLE
passando ihttpfactory vai DI para controller no teste

### DISCUSSION THREAD

**Discussion by LuizFredericoJr** [balta-io/7182#4] [Perguntas e Respostas] [answered]
Olá.
Como faço para passar o ihttpclientfactory para o meu controller, quando estou no meu projeto de testes??
Na API( **net core 3.1** ), o asp net core cuida da injeção de dependência e isso é transparente para o dev.
Mas o projeto de testes não faz isso por baixo dos panos, fica tudo por conta da gente...
Alguém pode me dar uma dica (prática, please) ?

exemplo do controler:
public class CashController : ControllerBase
    {
        public const string ControllerName = nameof(CashController);
        private readonly ILogger<CashController> _logger;
        private readonly IHttpClientFactory _clientFactory;
        private readonly IConfiguration _configuration;
        private readonly ILogger<OnTrackController> _loggerOnTrack;

        public CashInController(
                                                **IHttpClientFactory clientFactory**,
                                    ILogger<CashController> logger,
                                    ILogger<OnTrackController> loggerOnTrack,
                                    IConfiguration config)
        {
            _logger = logger;
            _clientFactory = clientFactory;
            _configuration = config;
            _loggerOnTrack = loggerOnTrack;
        }

**Comment by andrebaltieri** [comment, marked as answer]
@LuizFredericoJr o `ihttpclientfactory` é uma interface, você não consegue mockar ela?
PS:. Ter essa dependência nos testes pode ser ruim, não seria melhor criar uma camada (Um serviço talvez) para abstrair isto?

**Reply by LuizFredericoJr** [reply]
Ja tentei mockar, mas como depois ela é passada pra classe que fará o factory do client, teria que configurar(fazer um setup) isso e não sou craque no moq. Eu só preciso mesmo passar essa interface, pq vários controllers já estão assim.
Realmete preciso fazer os testes passando a interface.
Já vasculhei a internet e não achei nada que focasse apenas nisso.
Ou seja, qdo passar a interface pro controller, ela ja precisa estar 'injetada', pois o client só será criado em outra classe(que acessa config e recupera urls pra montar o client e fazer a chamada), que irá receber esta mesma interface. Pode não ser a solução mais adequada, mas agora realmente precisava fazer isso no projeto de testes.
Está valendo mstest, xunit ou qq outro, se funcionar vou fazer uma festa aqui kkk.
Obs.: já coloquei a referencia do Xunit.Microsoft.DependencyInjection pra ajudar...

**Reply by andrebaltieri** [reply]
Então, mas por que ao invés de depender do IHttp você não "esconde" ele atrás de um serviço e mocka este serviço nos testes?
A ideia é criar uma abstração para justamente tornar isto testável...

**Reply by LuizFredericoJr** [reply]
Balta, obrigado pelas dicas, sei que vc esta falando de padrões e boas práticas, mas as vezes não dá pra refatorar o projeto na hora em que gostaríamos, temos apenas de resolver o problema: (resolvido abaixo)

usings:
  using Microsoft.Extensions.Configuration;
  using Microsoft.Extensions.Logging;
  using System;
  using System.Net;
  using System.Net.Http;
  using System.Threading.Tasks;
  using System.Threading;
  using Xunit;
  using Microsoft.AspNetCore.Mvc;
  using System.IO;
  using Microsoft.Extensions.DependencyInjection;

No teste:
            var serviceCollection = new ServiceCollection();
            ConfigureServices(serviceCollection);
            var services = serviceCollection.BuildServiceProvider();
            var httpFactory = services.GetRequiredService<IHttpClientFactory>();


Metodo Auxiliar:
        private static void ConfigureServices(ServiceCollection services)
        {
            services.AddHttpClient();
        }

**Reply by andrebaltieri** [reply]
<3
